### DISCUSSION TITLE
Some issues with goccy/go-yaml

### DISCUSSION THREAD

**Discussion by SeWieland** [helmfile/helmfile#656] [General]
I would like to share some issues I discovered using the new "goccy/go-yaml" yaml parser that will be the default for V1.

I enabled it using helmfile v0.15.0 and setting `HELMFILE_GOCCY_GOYAML=true` because I use many anchors and aliases and got lots of duplicate key errors.

Sadly, the new yaml parses does not work for me "out of the box" - several issues occured.
I have some modules declared that I render dynamically using the `tpl` function. All these modules behave differently now.

## Example Setup:

I have the following definitions in my values for making my ingress anntotations more DRY, and they make use of some values as well:

```yaml
# The default ingress annotations
.x-default: &default
  nginx.ingress.kubernetes.io/auth-url: https://{{ .Values.authService }}
  nginx.ingress.kubernetes.io/auth-response-headers: Authorization
  acme.cert-manager.io/http01-edit-in-place: "true"
  nginx.ingress.kubernetes.io/ssl-redirect: "true"
  nginx.ingress.kubernetes.io/force-ssl-redirect: "true"
  cert-manager.io/cluster-issuer: letsencrypt-prod

# Additional "soft" stickyness for performance improvements whenever local session caching is involved 
.x-stickyness: &stickyness
  nginx.ingress.kubernetes.io/affinity: cookie
  nginx.ingress.kubernetes.io/affinity-mode: balanced
  nginx.ingress.kubernetes.io/session-cookie-name: _sticky_session
  nginx.ingress.kubernetes.io/session-cookie-change-on-failure: "true"
  
# Proxy caching can be used for caching static content in nginx
.x-proxy-caching: &caching
  nginx.ingress.kubernetes.io/proxy-buffering: "on"
  
###
# Ingress Annotation Templates
###

ingressAnnotations:
  default:
    << : *default
  caching:
    << : [*default, *caching]
  sticky:
    << : [*default, *stickyness]
```

I use them as follows in my yaml.gotmpl files:

```yaml
ingress:
  enabled: true
  annotations:
    {{ tpl ( .Values.ingressAnnotations.default | toYaml | nindent 4 ) . }}
  rules:
  - host: {{ .Values.domainName }}
    paths:
    {{- range .Values.service.ingressPaths }}
      - {{ . }}
    {{- end }}
  tls:
  - hosts:
    - {{ .Values.domainName }}
    secretName: tls-secret
```

### Merging multiple anchors

With `goccy/go-yaml` I cannot use `.Values.ingressAnnotations.caching` or  `.Values.ingressAnnotations.stickyness` anymore.
With array style anchor merging (`<< : [*default, *stickyness]`) the annotations are not rendered at all **without any error**.

This results in a valid ingress ressource without any annotations + only the diff shows that they are simply not rendered :/

I have to move back to the "old" `%YAML 1.1` declaration syntax (this works)

```yaml
ingressAnnotations:
  default:
    << : *default
  caching:
    << : *default
    << : *caching
  sticky:
    << : *default
    << : *stickyness
``` 

However..

### Some keywords are handeld differently

The merging works again now. But using `.Values.ingressAnnotations.caching` _still results in an error_.
`<$value>: wrong type for value; expected string; got bool`

The debug log shows what happened:
The rendered annotations:

```yaml 
annotations:
  nginx.ingress.kubernetes.io/auth-url: https://myAuthService
  nginx.ingress.kubernetes.io/auth-response-headers: Authorization
  cert-manager.io/cluster-issuer: letsencrypt-prod
  nginx.ingress.kubernetes.io/force-ssl-redirect: "true"
  nginx.ingress.kubernetes.io/proxy-buffering: on
  nginx.ingress.kubernetes.io/ssl-redirect: "true"
``` 

**The quotes of my "on" got stripped and I cannot bring them back**.

I have not yet found any way to work around this issue. The `tpl` function will remove the quotes and any effort to bring them back did not work yet.
But it led to yet another headache.

### Quotes won't work at all

I'm pretty desperate, so I tried the following:

```yaml
.x-proxy-caching: &caching
  nginx.ingress.kubernetes.io/proxy-buffering: '{{ quote "on" }}'
```

So that when I render the string through `tpl` it should just quote the given string, but nope.
I tested quite a bit and it turned out that helmfile will always exit with `error calling tpl: template: stringTemplate:16: unexpected "\\" in operand` for _any function that needs a string input_ in that constellation..
My templates for some common secrets like `'{{ printf "%s%s" .Values.keyVault "secret-ref" | fetchSecretValue }}'` are all broke with goccy/go-yaml..

## Conclusion

The switch to the new yaml parser will most certainly cause a lot of headaches in the community..
And I still do not know if I can use a lot of my "modules" I built for dynamic templating at all in future releases.

I guess the issues I discovered are not fixable in this project, and probably a won't fix in the goccy/go-yaml project.
However I hope these findings will help other users to make the switch to V1 a bit easier.

_Edit:_ minor clarifications

**Comment by yxxhero** [comment]
Please provide a example Project.

**Comment by yxxhero** [comment]
```
.x-proxy-caching: &caching
  nginx.ingress.kubernetes.io/proxy-buffering: !!str "on"
```

**Reply by SeWieland** [reply]
This did not work either => The debug output shows that `nginx.ingress.kubernetes.io/proxy-buffering: on` is rendered still.

**Comment by yxxhero** [comment]
@SeWieland I have known the reason. Please wait for me to complete it.

**Reply by yxxhero** [reply]
from the tests. not resolve the issue.

**Comment by yxxhero** [comment]
@SeWieland Thanks for your feedback very much.

**Comment by yxxhero** [comment]
https://github.com/goccy/go-yaml/issues/214

**Comment by yxxhero** [comment]
@mumoshu I think that this is a big question. What's your opinion?

**Comment by SeWieland** [comment]
So I've got a nasty workaround for the string vs bool issue..

Using `toJson` instead of `toYaml` will correctly produce `nginx.ingress.kubernetes.io/proxy-buffering: "on"`.
This could resolve many issues I guess, because JSON should work just fine..

**Comment by yxxhero** [comment]
@SeWieland because in Yaml spec 1.2. `on/off` is not bool. but `ToJson` uses Yaml Spec 1.1.

**Reply by SeWieland** [reply]
Thats kind of sad, because the YAML 1.1 style merge syntax works.. And my YAML linter will just be happy with it by adding
```
%YAML 1.1
---
```
to the start of the file... So it's even indicated that I want to use the 1.1 syntax in that particular file

**Reply by yxxhero** [reply]
I don't know the workaround. got it.

**Comment by yxxhero** [comment]
My templates for some common secrets like '{{ printf "%s%s" .Values.keyVault "secret-ref" | fetchSecretValue }}' are all broke with goccy/go-yaml..

I want to know more about the above issue.

@SeWieland

**Reply by SeWieland** [reply]
It's kind of a niche use-case (I already removed every occurance because it wasn't really needed after all), but the issue applies to any template you would want to render dynamically using `printf`.

=> E.g. declaring `'{{ printf "%s-%s" "example" "value" }}'` in one file and using it with `tpl` in another file will produce this error..

**Reply by SeWieland** [reply]
And my best guess is that the quotes `"` are handled incorrectly (escaped?) somehow..

**Comment by mumoshu** [comment]
@SeWieland Hey! Thank you so much for your detailed report. I'll definitely use your example setup as a reference to make the Helmfile v1 migration smooth as possible.

> With array style anchor merging (<< : [*default, *stickyness]) the annotations are not rendered at all without any error.

This sounds like a potential bug in goccy/go-yaml. Is this supposed to work in any YAML 1.2 compliant parser?
If so, we might submit a bug report to goccy/go-yaml and submit a pull request to fix it.

> The merging works again now. But using .Values.ingressAnnotations.caching still results in an error.
> <$value>: wrong type for value; expected string; got bool

Who's producing the error message? Maybe Helm? If so, that makes sense. Helm might be using gopkg.in/yaml.v2 or v3, which is seemingly not compliant with YAML 1.2(That's the possible reason the parser isn't treating `on` as a regular string, right?)

Anyway, regardless of which yaml parser Helmfile uses, it needs to produce YAML documents that can be intepreted by whatever YAML parser used by the downstream products and tools, like kubectl and helm.

So if we are to go with goccy/go-yaml, we shall submit a feature request to goccy/go-yaml to let it "force quote" any strings, so that `on` is encoded as `"on"` by goccy/go-yaml.

WDYT?

**Reply by mumoshu** [reply]
TL;DR; we shall submit one bug report and one feature request to goccy/go-yaml, submit pull requests for those, upgrade goccy/go-yaml in helmfile, and your example would work out-of-box. Does it sound good to you @SeWieland?

**Reply by SeWieland** [reply]
That sounds quite perfect @mumoshu 👌 
I haven't read the YAML spec, but the array style merging approach is pretty common practice since YAML 1.2, I think.

Your assumption that helm produces the "wrong type for value" error is produced by helm should be correct, as I already saw this error on "helm only" deployments as well.

**Comment by mumoshu** [comment]
@SeWieland @yxxhero TIL that YAML 1.2.2 does not seem to define "<<: [*foo, *bar]" syntax in https://yaml.org/spec/1.2.2/#anchors-and-aliases and https://yaml.org/spec/1.2.2/#71-alias-nodes. Perhaps it's only present in older versions of YAML? 🤔 

Can you instead migrate to the new `inherit` feature that we recently introduced via #606? It doesn't support multi-parents like `<< : [*default, *stickyness]`. If that's a necessity, we might be able to enhance the `inherit` feature to optionally support multiple parents.

**Reply by mumoshu** [reply]
Apparently it's an optional spec and has never been defined or proposed for YAML 1.2 yet? 🤔 
https://ktomk.github.io/writing/yaml-anchor-alias-and-merge-key.html

**Reply by mumoshu** [reply]
> Merge Keys are only part of YAML 1.1 which is deprecated, not part of YAML 1.2 nor 1.0. That means that Merge Keys are born deprecated.

😞

**Reply by yxxhero** [reply]
oh. got it.

**Reply by mumoshu** [reply]
Here's the pull request to fix the `inherit` feature to fully replace the merge-key feature: #674

**Reply by SeWieland** [reply]
> > Merge Keys are only part of YAML 1.1 which is deprecated, not part of YAML 1.2 nor 1.0. That means that Merge Keys are born deprecated.
> 
> 😞

Wow. This will break so much stuff worldwide :D

I'm already switiching to the great new `inherit` feature in all my release specs 👍 
I think this is sufficient and the readability is better anyways.
